### DISCUSSION TITLE
slskd-python-api

### DISCUSSION THREAD

**Discussion by bigoulours** [slskd/slskd#910] [Show and tell]
Hey!

First, thank you for the great work you're doing with slskd!

Seeing you have been working on [Lidarr integration](https://github.com/slskd/slskd/issues/451) for a while, I wondered if it wouldn't be easier/faster to integrate with [headphones](https://github.com/rembo10/headphones) first. (I work mostly with python so I wouldn't be of big use with C#).
So I started to look into the API by activating the swagger endpoint. Figuring out how everything works kept me busy for a while and generating a client with OpenAPI/Swagger didn't yield anything useful. So in order to save other potential python developers the hassle, I wrote a python wrapper: [slskd-python-api](https://github.com/bigoulours/slskd-python-api).

Thanks to the great tools available (readthedocs, pypi, github, actions...) the generation of a user-friendly [documentation](https://slskd-api.readthedocs.io/en/latest/index.html) was relatively easy. I hope this will be helpful for others (not only python devs).

That being said, I have a few remaining questions about the api:

- I couldn't get `/api/v0/application/dump` to work. Always getting error 'Could not find file /tmp/...' is there some parameter to change to make it work?
- '/api/v0/users/{username}/browse/status' keeps returning Error 404 'Not Found'. Is it actually implemented?
- I guess you are working on some changes to the API for the Lidarr integration. Will the next API be backward compatible? 
- I haven't tested the PublicChat and Relay API since I have no use for them. If someone reading this would like to test them, feedbacks would be more than welcome!
- what is the purpose of passing a token when starting a search?

Cheers,

**Comment by jpdillingham** [comment]
The `/dump` route was an experimental attempt to make the application generate a heap dump and send it to the calling user.  It only works on Windows and shouldn't really be considered part of the external API.

The `/browse/status` route will only return information while a browse operation is running.  It's useful to check the status of the download.

The API is at `v0` right now to indicate that breaking changes can be made, but that's not spelled out anywhere.  I plan to move to v1 after I've had a chance to go over the routes and ensure consistency, etc.  For example, any route that uses a username is going to be refactored somehow, potentially to use base62 encoding or some other method; forward slashes in usernames breaks a bunch of stuff.

**Reply by bigoulours** [reply]
Thanks for the reply! I will leave `/dump` commented then. As for `browse/status`, it is probably best to call it asynchronously while `/browse` is still running. This is why I couldn't get anything else than error 404 before (now I was able to get a result).

Would you mind explaining tokens for searches? I still can't figure out why both token and uuid are needed.

Changing usernames endpoints in future versions of the API shouldn't be a problem. I will follow your Major version number to avoid confusion.

Also one last thing: according to the Swagger UI, searchTimeout is supposed to be in s, but after testing it looks like it should be set in ms.

**Reply by jpdillingham** [reply]
Sorry I neglected to answer that one.  Soulseek uses an integer as a search token under the hood, and the token in the POST body is meant to represent that, in case a caller cared what it was.  It's actually unused right now and I'll be removing it later.  The GUID/UUID is what's used to track the search in slskd and Soulseek.NET (the client under the hood) manages the integer based token automatically.

Note that the major version is still 0 and I'm not following SemVer rules at the moment.  Once I hit v1 I'll start versioning properly.

If you're planning to sink a bunch of time into this I can prioritize a cleanup pass on the API to help you minimize rework, just let me know.

**Reply by bigoulours** [reply]
Ok thanks.
I'll release a 0.1 version with a few pending changes and won't touch it unless bugs are found or you update API v0.
Once you release v1 I'll rework it. No rush though, summer has started so I won't spend too much time on it in the next months. Moreover, v0 is already enough to tinker with slskd.

**Comment by AdeHub** [comment]
Hi

Headphones now has provisional support for slskd in the develop branch

Thanks

**Comment by bigoulours** [comment]
Hi @jpdillingham !

I updated to v0.23 and noticed a few changes to the API. So here are my questions about them:

1.  the metrics can now be accessed at the same base url as the rest. Meaning there are two different ways to get them. Will both be kept or is one to be deprecated at some point?
2. i'm able to raise events but passing a string to "disambiguate generated events" doesn't seem to work. Does the string have to be in a specific format in order to be parsed? (tried json wiithout success, maybe some keys missing?)

Thanks and keep up the good work!

**Comment by bigoulours** [comment]
@jpdillingham I wrote my message shortly before the release of 0.24.0 and saw that you moved the metrics under /telemetry.
Will the old metrics endpoint remain?
Also still not luck with event string for disambiguation, do you have any idea what I might be doing wrong?
Thanks and regards,
